src/network/operator.rs

```diff
@@ -0,0 +1,336 @@
+    /// The peers router of the node.
+    peers_router: PeersRouter<N, E>,
+    /// The ledger state of the node.
+    ledger_reader: LedgerReader<N>,
+    /// The ledger router of the node.
+    ledger_router: LedgerRouter<N>,
+    /// The prover router of the node.
+    prover_router: ProverRouter<N>,
+}
+
+impl<N: Network, E: Environment> Operator<N, E> {
+    /// Initializes a new instance of the operator.
+    pub async fn open<S: Storage, P: AsRef<Path> + Copy>(
+        tasks: &Tasks<JoinHandle<()>>,
+        path: P,
+        address: Option<Address<N>>,
+        local_ip: SocketAddr,
+        memory_pool: Arc<RwLock<MemoryPool<N>>>,
+        peers_router: PeersRouter<N, E>,
+        ledger_reader: LedgerReader<N>,
+        ledger_router: LedgerRouter<N>,
+        prover_router: ProverRouter<N>,
+    ) -> Result<Arc<Self>> {
+        // Initialize an mpsc channel for sending requests to the `Operator` struct.
+        let (operator_router, mut operator_handler) = mpsc::channel(1024);
+
+        // Initialize the operator.
+        let operator = Arc::new(Self {
+            address,
+            local_ip,
+            state: Arc::new(OperatorState::open_writer::<S, P>(path)?),
+            block_template: RwLock::new(None),
+            provers: Default::default(),
+            operator_router,
+            memory_pool,
+            peers_router,
+            ledger_reader,
+            ledger_router,
+            prover_router,
+        });
+
+        if E::NODE_TYPE == NodeType::Operator {
+            // Initialize the handler for the operator.
+            let operator_clone = operator.clone();
+            let (router, handler) = oneshot::channel();
+            tasks.append(task::spawn(async move {
+                // Notify the outer function that the task is ready.
+                let _ = router.send(());
+                // Asynchronously wait for a operator request.
+                while let Some(request) = operator_handler.recv().await {
+                    operator_clone.update(request).await;
+                }
+            }));
+            // Wait until the operator handler is ready.
+            let _ = handler.await;
+        }
+
+        if E::NODE_TYPE == NodeType::Operator {
+            if let Some(recipient) = operator.address {
+                // Initialize an update loop for the block template.
+                let operator = operator.clone();
+                let (router, handler) = oneshot::channel();
+                tasks.append(task::spawn(async move {
+                    // Notify the outer function that the task is ready.
+                    let _ = router.send(());
+                    // TODO (julesdesmit): Add logic to the loop to retarget share difficulty.
+                    loop {
+                        // Determine if the current block template is stale.
+                        let is_template_stale = match &*operator.block_template.read().await {
```

Comment:
Is a template ever stale other than the initial start?

Reply:
The template should go stale ever 20 seconds